### ISSUE TITLE
adnuts considerations for 2022

### ISSUE THREAD

**Issue by aaronmberger-nwfsc** [pacific-hake/hake-assessment#835]
A couple things to consider:

1. ***important: we should switch back to the most current version of adnuts for next year (main branch, rather than the dev branch used in 2021)
2. We should be setting seeds= in the sample_xxxx functions, for reproducibility (at least when using the same number of cores); you can check the seed by using nuts.updated$cmd
3. We should run exploratory models with adapt= 0.90, but up it to adapt=0.99 when running the final determined base model as the diagnostics are better (run time is a bit longer). We should look at how much longer this takes and consider if a higher adapt should be used for sensitivities and retrospectives.
4. You can add a duration=XX to the sample_xxxx functions to allow a maximum number of minutes you will allow the algorithm to run (i.e., achieving a set number of samples to end the algorithm is overridden by a time constraint if it times out first) - this is good for explorations but not something to be used on final runs.  Although might be one way to limit those super long retro years.

**Comment by cgrandin** [comment]
Regarding number 3 above, do we use 0.99 for all sensitivities and retrospectives as well? I'm wondering how much time that would add to our computations.

**Comment by aaronmberger-nwfsc** [comment]
That is something we will have to test, b/c it inevitably will add time.  My thoughts as of today is that the adapt=0.90 is good enough for sensitivities/etc, especially when diagnostics don't show patterns with divergences. It is just the base model used for management that we should invest the additional time to run the adapt=0.99.  I should have tracked the time difference this year, but I didn't - though I don't recall being surprised.

**Comment by aaronmberger-nwfsc** [comment]
Comparing performance of identical models evaluated with adnuts using 1) adapt_delta = 0.9, 2) adapt_delta = 0.95, and 3) adapt_delta = 0.99.   

**Computing time:**
Generated ~8000 posterior samples without running retrospectives and forecasts
adapt_delta = 0.90 took about 1.5 hours on average
adapt_delta = 0.95 took only a bit longer (my comparison wasn't fully apples to apples, but the additional time was not overwhelming
adapt_delta = 0.99 took almost 11 hours, so considerably increase

**Diagnostics:**
Overall, performance diagnostics improved with increasing the target acceptance rate (adapt_delta)
![image](https://user-images.githubusercontent.com/7946234/148429640-c9242267-7dc3-465f-828b-692b2592ef92.png)

**Management Quantities:**
Overall, little change in quantities directly used for management, but some small differences
![image](https://user-images.githubusercontent.com/7946234/148429795-7fb4d722-21d8-4aca-a4ea-72482ca050eb.png)

Differences in depletion-based stock status (Bratio) are mostly within 2 significant digit rounding error:
![image](https://user-images.githubusercontent.com/7946234/148436645-d7bda1a5-5f35-4153-8a61-917cbec399ad.png)

My thoughts on this are that we can run the base model with adapt_delta = 0.99 to essentially show 'extra good' diagnostics, but it is not really worth the added computational burden in terms of model results used for management.  So I think we are perfectly able to use adapt_delta = 0.95 for the base model, and lose very little.  Moving to adapt_delta=0.90 is less desirable (at least for the base model) due to the increasing divergent error rate and the decent amount of effective sample size loss for key parameters (e.g., M and R0).  A divergent error rate of 2.5% (205 out of 8250) was noted as something to reduce when I discussed this last year with @Cole-Monnahan-NOAA.  Interestingly, some rec devs had a higher relative effective sample size when using adapt_delta=0.90, but I think that is more likely due to the fact that I ran that model with 11 cores/chains; whereas the adapt_delta = 0.95 and 0.99 models were run on my remote desktop with 7 cores/chains.  So maybe there is a core/chain effect too.  But all and all that is pretty minimal. 

In summary, I suggest:
1.  Use of adapt_delta=0.95 for base model
2.  Use of adapt_delta=0.90 for sensitivities and retrospectives
3.  Adjust these suggestions upwards as needed depending visual diagnostics via launch_shinyadmb, such that these are minimums.

**Comment by Cole-Monnahan-NOAA** [comment]
It's really promising to see that your divergences go down with more accurate trajectories, and that the overall inference is really stable. You want to look at minESS not maxESS. Plus there's just a lot of uncertainty about those ESS estimates so it's not a surprise to see them wiggling around even with such high posterior samples.

One suggestion for shorter runs in sensitivities would be to run shorter chains. What are your inputs? If you use the metric='mle' option and start from the mode then the warmup can be quite short. It just needs a few iterations to get to the typical set and then maybe 50-100 to estimate a good step size. There's a function `adnuts::plot_sampler_params` which shows what the tuning is doing FYI.

It's fun to see NUTS out there in the wild.

**Comment by kellijohnson-NOAA** [comment]
Thanks @Cole-Monnahan-NOAA for all of the development on {adnuts} and for always chiming in when we tag you. Your insight is super helpful!

**Comment by aaronmberger-nwfsc** [comment]
As always, we really appreciate your insights @Cole-Monnahan-NOAA.  Your idea for shortening sensitivity runs sounds promising.  We are currently using metric='mass' for sensitivity runs as well, where the inits are from a shorter chain that does use the metric='mle' and a target acceptance rate of 0.8 to get us in the ballpark. 

It would be interesting to have a play with optimizing sensitivity runtimes across the range of sensitivity models (i.e., from small one-off single parameters changes to large data/structural changes).

Our biggest advantage is having multi-thread/core laptops, 'server' style computing access, and cloud computing so that we can farm out all models at, or nearly at, once.

**Comment by Cole-Monnahan-NOAA** [comment]
Yeah I'd definitely play around with minimum settings to get reasonable convergence. You can plot things like cumulative metrics (over randomized posteriors) to see when it stabilizes. That'll give you a sense of the minimum number of samples. If your warmup is too short the adapted step sizes will be really different among chains and potentially biased low due to starting at the mode (easiest place to sample from). Again, look at the sampler_param plot to get a sense of the behavior of eps during warmup.

**Comment by andrew-edwards** [comment]
Awesome analyses. @aaronmberger-nwfsc - you said:
2. Use of adapt_delta=0.90 for sensitivities and retrospectives

But why not just stick with 0.95 for those if that's going to be used for the base case, and didn't seem to take any longer? 

Or maybe 0.95 for sensitivities (which I think generally take as long as the base case). And I just realised that perhaps 0.90 is indeed best for retrospectives since some of those took ages last year. 

I'm good either way, just curious.

**Comment by aaronmberger-nwfsc** [comment]
I was thinking of minimums, but yes we certainly could increase to adapt_delta=0.95 for sensitivities if run time isn't too long and divergence errors are high (or other less desirable diagnostics).  I am fine starting there and then reducing only if needed.  That way our code can stay with 0.95 and not need to be adjusted depending on the model run (unless the analysts deems it necessary when looking at resulting diagnostics).

**Comment by kellijohnson-NOAA** [comment]
Files are set and running with 0.95.
